src/renderer/entities/transaction/lib/transactionService.ts

```diff
@@ -28,21 +28,107 @@ export type TxWrappers = WrapAsMulti | WrapAsProxy;
 const shouldWrapAsMulti = (wrapper: TxWrappers): wrapper is WrapAsMulti =>
   'signatoryId' in wrapper && 'account' in wrapper;
 
+export const createRegistry = async (api: ApiPromise) => {
+  const metadataRpc = await api.rpc.state.getMetadata();
+
+  return {
+    registry: api.registry as unknown as TypeRegistry,
+    metadataRpc: metadataRpc.toHex(),
+  };
+};
+
+export const getSignedExtrinsic = async (
+  unsigned: UnsignedTransaction,
+  signature: HexString,
+  api: ApiPromise,
+): Promise<string> => {
+  const { metadataRpc, registry } = await createRegistry(api);
+
+  return construct.signedTx(unsigned, signature, { registry, metadataRpc });
+};
+
+export const submitAndWatchExtrinsic = (
+  tx: string,
+  unsigned: UnsignedTransaction,
+  api: ApiPromise,
+  callback: (executed: boolean, params: ExtrinsicResultParams | string) => void,
+) => {
+  let extrinsicCalls = 0;
+
+  const callIndex = api.createType('Call', unsigned.method).callIndex;
+  const { method, section } = api.registry.findMetaCall(callIndex);
+
+  api.rpc.author
+    .submitAndWatchExtrinsic(tx, async (result) => {
```

Comment:
Can you tell me more about the reason behind using `.submitAndWatch`? Knowing very little about how Walletconnect would handle that, I'd have expected that we could use `.send` that then gives the Tx events etc.

Reply:
WalletConnect signing is not connected with `submitAndWatchExtrinsic`
`submitAndWatchExtrinsic` happens after signing was already made

Reply:
Ok, so do you know why we're using `submitAndWatchExtrinsic` rather than [`.send`](https://github.com/polkadot-js/api/blob/390610127a350a8ebb5bac2b8c6645c50060a7c6/packages/api-base/src/types/submittable.ts#L65)? the send allows to pass a callback, and will give you all the status, events right away, without the need to subscribe.

Reply:
@Asmadek need your help with Thibaut's question

Reply:
The idea is that you woulnd't need to do all the following
```
      const signedBlock = await api.rpc.chain.getBlock();
      const blockHeight = signedBlock.block.header.number.toNumber();
      const apiAt = await api.at(signedBlock.block.header.hash);
      const allRecords = await apiAt.query.system.events();
```

But rather directly access the status and event as explained [here](https://polkadot.js.org/docs/api/start/api.tx.subs#transaction-events), or [here](https://polkadot.js.org/docs/api/cookbook/tx#how-do-i-get-the-decoded-enum-for-an-extrinsicfailed-event) e.g:

```
.send(sender, ({ status, events }) => {
    if (status.isInBlock || status.isFinalized) {
      events
        // find/filter for failed events
        .filter(({ event }) =>
          api.events.system.ExtrinsicFailed.is(event)
        )
        // we know that data for system.ExtrinsicFailed is
        // (DispatchError, DispatchInfo)
        .forEach(({ event: { data: [error, info] } }) => {
          if (error.isModule) {
            // for module errors, we have the section indexed, lookup
            const decoded = api.registry.findMetaError(error.asModule);
            const { docs, method, section } = decoded;

            console.log(`${section}.${method}: ${docs.join(' ')}`);
          } else {
            // Other, CannotLookup, BadOrigin, no extra info
            console.log(error.toString());
          }
        });
    }
```

Reply:
@Asmadek @Tbaut I agree, will create task for it
Thanks for such a valuable feedback!
update: [task](https://github.com/novasamatech/nova-spektr/issues/1585)